javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
I've made some tests to verify whether it's actually required, in practice, to have a field marked as volatile when used in a double-checked lock. Theoretically, the mechanism can fail without the field being marked as volatile. 

According to my tests, it turns out that this is very much JVM-dependent. My test never failed when running against a Java 7 and 8 JVM. However, it did fail against Java 6.

My conclusion is that, at least since Java 7, there are some different runtime optimizations in place that either decrease the chance of the problem happening or make the _volatile_ modifier completely unneeded in this double-checked locking scenario. However, we should definitely not rely on this apparent behavior until there is an update to the Java Memory Model that clearly states that we can do so. 

I'll give you more details as soon as I got the time.